internal/controller/pkg/resolver/reconciler.go

```diff
@@ -375,3 +401,95 @@ func (r *Reconciler) findDependencyVersion(ctx context.Context, dep *v1beta1.Dep
 
 	return addVer, nil
 }
+
+// UpdatableVersionFinder is an implementation of versionFinder that allows for version upgrade/downgrade capability considering parent constraints.
+type UpdatableVersionFinder struct {
+	fetcher xpkg.Fetcher
+	config  xpkg.ConfigStore
+}
+
+// FindValidDependencyVersion finds a valid version with version upgrade/downgrade capability considering parent constraints.
+func (u *UpdatableVersionFinder) FindValidDependencyVersion(ctx context.Context, dep *v1beta1.Dependency, ref name.Reference, n dag.Node, log logging.Logger) (string, error) {
+	if digest, err := conregv1.NewHash(dep.Constraints); err == nil {
```

Comment:
Not introduced with this PR but why it is `dep.Constraints` given it is not a list but only a string ? Why not `dep.Constraint` ?